### ISSUE TITLE
Design - register aio handlers for events

### ISSUE THREAD

**Issue by dazza-codes** [aio-libs/aiobotocore#761]
Botocore issues many event triggers - could aiobotocore operate like a plugin to register handlers with a botocore session?

- https://botocore.amazonaws.com/v1/documentation/api/latest/topics/events.html
  - `before-call` might substitute an aiohttp client

It might be better if the botocore session could register a component for a client creator or the client creator could register a client component or the `BaseClient` could register a component to make-request that supports asyncio coroutines.  This is not yet possible without a PR on botocore that might or might not be reviewed and merged.

Could a simple PR on botocore try to add a `botocore.config.Config` option for something like `use_aiohttp` that might switch things up when it creates a client to return aio-clients?  Could this add a new base-class that could be used instead of the `BaseClient`, it seems like this is what aiobotocore is doing with `AioBaseClient`, but it has to copy/maintain a ton of botocore apparatus.

**Comment by thehesiod** [comment]
tying to get anything merged into botocore is a disaster, I've had PRs open for years that don't get merged or commented on.  Good luck with that :)

**Comment by thehesiod** [comment]
plus I don't think it'll work because their functions need to be async and they're not in botocore.  It can call our event handler and then we return some coroutine but it can't await that in the middle of it's function.  I'm going to close this for now, if you see a concrete way to do this feel free to re-open.  The right way to do this is everything (botocore + aiobotocore) to switch to use their c++ sdk https://github.com/aws/aws-sdk-cpp because  botocore is very slow.  It can only do ~50 requests in parallel per process due to all the overhead and inefficiencies with doing this in python (and that's with a patch that they never merged of mine)

**Comment by thehesiod** [comment]
also feel free to expand on https://github.com/aio-libs/aiobotocore/issues/36
